### ISSUE TITLE
Firebase

### ISSUE THREAD

**Issue by miickel** [vandium-io/vandium-node#10]
Has anyone used vandium successfully together with Firebase? My firebase lambda function works fine without vandium, but with the plugin activated (barebone setup) the task times out.

I have narrowed down the cause of the timeout to this line:

`firebase.database().ref('stuff')`

But I'm not sure what might cause the timeouts. A wild guess would be that this plugin's promises implementation might crash together with firebase's.

Any ideas what might cause this behaviour?

Thanks

**Comment by richardhyatt** [comment]
The database has a socket open which prevents the lambda from exiting (this is not a Vandium thing but how lambda works with the callback function). We had a similar issue with Redis that caused the lambda function to timeout and we replicated it using a pure nodejs implementation. 

Either you close the connection at the end of your handler using Firebase's [`goOffline()`](https://www.firebase.com/docs/web/api/firebase/gooffline.html) function or use Vandium's [`after()`](https://github.com/vandium-io/vandium-node/blob/master/docs/cleanup.md) function to  call `goOffline()` to free the resources. The `after()` function will get invoked on success or failure.

Also, if you want to verify if resources are being leaked, you can use our [`lambda-tester`](https://github.com/vandium-io/lambda-tester) package - it has an [option](https://github.com/vandium-io/lambda-tester/blob/master/docs/main.md#resource-leak-detection) that you can enable to detect leaks.

Let me know if this fixes the issue.

**Comment by miickel** [comment]
I tried using `after()` and also calling `goOffline()` but my function still timeout. Note: it only timeout when adding vandium to the handler.

Here's a minimal example that replicates the error:

```javascript
const Firebase = require('firebase');
const firebase = Firebase.initializeApp({
  // config
});

module.exports.handler = handler; // wrapping this with vandium() causes timeout

function handler(event, context, callback) {
  context.callbackWaitsForEmptyEventLoop = false;
  firebase.database()
    .ref('stuff')
    .once('value')
    .then(() => {
      callback(null, 'ok');
    });
}
```

Not wrapping the handler in vandium() returns 'ok', wrapping the handler in vandium() timeout (6 seconds).

Thanks!

**Comment by dc00p** [comment]
Calling `goOffline()` does not actually cause the Firebase module to fully stop executing, and as a result, your lambda function continues to run, until it times out.

See http://stackoverflow.com/questions/38560186/process-timeout-amazon-lambda-to-firebase for more details.

We are currently investigating integrating the solutions provided in the above article into vandium to handle cases such as yours.

**Comment by richardhyatt** [comment]
Doesn't look like firebase is playing nice, so we've support for `context.callbackWaitsForEmptyEventLoop = false` in version 3.2.

Let me know if this resolves the issue.

**Comment by miickel** [comment]
Great!

3.2 works with a no-config setup of vandium just wrapping my handler, but when I add some validation the task times out again (btw, there seems to be some ambiguity in docs and examples here, I've tried both methods):

```
vandium.validation({
  idToken: 'string:min=1,trim,required',
  registrationId: 'string:min=1,trim,required',
});
```

**Comment by richardhyatt** [comment]
Try the [3.x](https://github.com/vandium-io/vandium-node/tree/3.x) branch to confirm that the changes work for you - it will be the 3.3.0 release. You can install into your project via `npm install vandium-io/vandium-node#3.x` and use the new configuration option `vandium. callbackWaitsForEmptyEventLoop( false )`.

**Comment by richardhyatt** [comment]
Resolved in 3.3.0
